### PR TITLE
Make most emojis full-width

### LINKED ISSUE #900: Certain "emoji" are still half-sized
# Environment

```none
Windows build number: 10.0.18899.1000
Windows Terminal version: 0.1.1361.0
```
# Steps to reproduce

Setup a standard PowerLine using FiraCode in WSL(2) and navigate to a git repository that contains edits. Note if you 
`printf '✏' ` that the pencil is double the size of the one in the Powerline prompt. 

image

# Expected behavior

I expect emojis in the prompt to appear at the same size as those later in the same line.

# Actual behavior

emoji size appears to be halved? See pic.

### PR DESCRIPTION
**Pull request by leonMSFT**
The table that we refer to in `CodepointWidthDetector.cpp` to determine
whether or not a codepoint should be rendered as Wide vs Narrow was
based off EastAsianWidth[1].  If a codepoint wasn't included in this
table, they're considered Narrow. Many emojis aren't specified in the
EAW list, so this PR supplements our table with emoji codepoints from
emoji-data[2] in order to render most, if not all, emojis as full-width. 

There are certain codepoints I've added to the comments (in case we want
to add them officially to the table in the future) that Microsoft
decided to give an emoji presentation even if it's specified as
Narrow/Ambiguous in the EAW list and are _not_ specified in the Unicode
emoji list. These include all of the Mahjong Tiles block, different
direction pencils (✎✐), different pointing index fingers (☜, ☞) among
others. I have no idea if I've captured all of them, as I don't know of
an easy way to detect which are Microsoft specific emojis.

## Validation Steps Performed
I have looked at so many emojis that I dream emoji.

These screenshots aren't encompassing _all_ emoji but I've tried to grab
a couple from all across the codepoint ranges:

Before:
before

After:
after

[1] http://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt
[2] https://www.unicode.org/Public/13.0.0/ucd/emoji/emoji-data.txt

Closes #900

### CODE REVIEW

**Comment by zadjii-msft** [comment]
@msftbot make sure @miniksa signs off on this

File: src/types/CodepointWidthDetector.cpp [hunk 1]
```diff
@@ -147,17 +156,21 @@ namespace
         UnicodeRange{ 0x22bf, 0x22bf, CodepointWidth::Ambiguous },
         UnicodeRange{ 0x2312, 0x2312, CodepointWidth::Ambiguous },
         UnicodeRange{ 0x231a, 0x231b, CodepointWidth::Wide },
-        UnicodeRange{ 0x2329, 0x232a, CodepointWidth::Wide },
-        UnicodeRange{ 0x23e9, 0x23ec, CodepointWidth::Wide },
-        UnicodeRange{ 0x23f0, 0x23f0, CodepointWidth::Wide },
-        UnicodeRange{ 0x23f3, 0x23f3, CodepointWidth::Wide },
-        UnicodeRange{ 0x2460, 0x24e9, CodepointWidth::Ambiguous },
+        UnicodeRange{ 0x2328, 0x232a, CodepointWidth::Wide },
```
Comment (DHowett-MSFT) [inline src/types/CodepointWidthDetector.cpp:159]: If possible -- and I know this is a lot of work -- i'd like a comment on each line that refers to a footnote about us overriding these. Just so we can come find it later.

**Comment by j4james** [comment]
I'm not an expert on this, but I've followed a few other discussions on this subject, and my understanding was that it was vital that we exactly follow whatever standard everyone else uses for determining character width. Applications like editors need to be able to predicate exactly how much of a line of text they can fit on the screen before it wraps. They can't do that if the standard says a character should be narrow, but we're rendering it as wide.

I don't know exactly what the standards are that we should be following, so maybe everything in this PR is good, but the mention of "Microsoft specific emojis" makes it sound like we're making up own rules for some characters, and I suspect that's going to cause problems.

**Comment by miniksa** [comment]
> I'm not an expert on this, but I've followed a few other discussions on this subject, and my understanding was that it was vital that we exactly follow whatever standard everyone else uses for determining character width. Applications like editors need to be able to predicate exactly how much of a line of text they can fit on the screen before it wraps. They can't do that if the standard says a character should be narrow, but we're rendering it as wide.
> 
> I don't know exactly what the standards are that we should be following, so maybe everything in this PR is good, but the mention of "Microsoft specific emojis" makes it sound like we're making up own rules for some characters, and I suspect that's going to cause problems.

I agree with James here. If Microsoft are the only one emojifying a particular Unicode codepoint, we should just center that glyph inside the 1 cell that it's supposed to be taking up.

**Comment by leonMSFT** [comment]
That's fair, I was unsure about putting the Microsoft specific emojis in, because if I didn't, the `lower_right_pencil` `u270e` emoji that was noticed to be tiny in #900 would stay tiny since it isn't in the standard.

**Comment by miniksa** [comment]
> That's fair, I was unsure about putting the Microsoft specific emojis in, because if I didn't, the `lower_right_pencil` `u270e` emoji that was noticed to be tiny in #900 would stay tiny since it isn't in the standard.

Yeah that's why I said maybe we make it an experimental toggle that is off by default. So purists can have what Unicode dictates and Microsoft fanboys can have their full width pencils.

**Comment by DHowett-MSFT** [comment]
> Yeah that's why I said maybe we make it an experimental toggle that is off by default. So purists can have what Unicode dictates and Microsoft fanboys can have their full width pencils.

I am deeply concerned with us plugging a configuration option into _this global singleton function_ (since right now we don't have a good place to store state fo codepoint width detection, and we _need_ conpty to agree on the widths with terminal, and there's no good way to communicate that state down to conpty)

**Comment by leonMSFT** [comment]
I'll leave the MS codepoints I've found in the comments so we can come back to implement them as we wish in the future then?

**Comment by miniksa** [comment]
> > Yeah that's why I said maybe we make it an experimental toggle that is off by default. So purists can have what Unicode dictates and Microsoft fanboys can have their full width pencils.
> 
> I am deeply concerned with us plugging a configuration option into _this global singleton function_ (since right now we don't have a good place to store state fo codepoint width detection, and we _need_ conpty to agree on the widths with terminal, and there's no good way to communicate that state down to conpty)

Oof. OK. Yeah you're right.



> I'll leave the MS codepoints I've found in the comments so we can come back to implement them as we wish in the future then?

Yeah that's probably best.

**Comment by DHowett-MSFT** [comment]
James is right, but right now the best thing we have is "vague consensus of various `wcwidth` implementations" and we're only making things _slightly more inconsistent_ by adding another one. I'm not happy that we have to do this, but I think that applications are going to have a rough time _regardless_ since the landscape here is so bad. `wcwidth` operates on a `wchar_t`, and there's no practical way to handle things like the emoji variation selector or other composing characters.

Unfortunately, emoji are just one place where column-counting breaks down for applications. ☹️

**Comment by TapeWerm** [comment]
When I type a ☢ emoji and backspace it in Ubuntu 18.04 I need to Ctrl-L to make it disappear. This does not happen on cmd. Should I open a separate issue for this?

**Comment by DHowett-MSFT** [comment]
Nope. We fixed this with #5934 :smile: we got some of the widths wrong

**Pull request merged**